handler/src/main/java/io/netty/handler/ssl/SslHandler.java

```diff
@@ -641,7 +646,14 @@ private void wrapNonAppData(ChannelHandlerContext ctx, boolean inUnwrap) throws
                 }
             }
         } catch (SSLException e) {
-            setHandshakeFailure(ctx, e);
+            try {
+                setHandshakeFailure(ctx, e);
+            } finally {
+                // We may have written some parts of data before an exception was thrown so ensure we always flush.
+                // See https://github.com/netty/netty/issues/3900#issuecomment-172481830
+                mayFlush(ctx);
```

Comment:
`handshake` (catches a SSLException and does a flush) -> `wrapNonAppData`
`unwrap` (catches a SSLException but does **not** do a flush) -> `wrapNonAppData`.

For this particular case should we just handle the flushing in `unwrap`'s exception catching code?

Seems like we may have some duplication in exception handling. The code paths also become a bit convoluted as `unwrap` may call `wrap`. I have not done analysis but my concern is there may be another code path that could also write an "error"/"ssl alert" but not be flushed (let me know if you can confirm this is not the case).